### PR TITLE
Avoid top-level `with ...;` in `nixos/doc/`

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

Using `with` at a top-level scope is [an anti-pattern](https://nix.dev/guides/best-practices#with-scopes). The tracking issue is #208242. This is a pure refactor: there is no functional change contained in this PR.

Spot the bug with `drvPath = guard drvPath;` though. Never evaled, and under a global `with`, it's not an error.

I used the refactor strategy that looks for the uses of names [coming from `lib`](https://gist.github.com/philiptaron/04326000d5ff20cb72c2805efd09701c). 

Following advice, I also preferred to inherit names from `lib` instead of `builtins` where there was a choice. I also preferred to inherit from `lib` in one place, rather than to have a mixture of `with lib;`, `lib.thing` and `inherit (lib) thing`. **I can drop those changes if reviewers desire.**

There are no rebuilds when I run `nixpkgs-review rev HEAD`, so I've targeted master with this PR.

## Things done

- [x] Tested compilation of all packages that depend on this change using `nixpkgs-review rev HEAD`.
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/misc/documentation.nix [hunk 1]
```diff
@@ -176,7 +200,7 @@ in
       enable = mkOption {
         type = types.bool;
         default = true;
-        description = lib.mdDoc ''
+        description = mdDoc ''
```
Comment (hsjobeki) [inline nixos/modules/misc/documentation.nix:203]: @pennae @infinisil; I guess these `mdDoc` markers are not needed anymore?
Since @philiptaron is cleaning the `with` statements we could also remove this unneded filewide "global"

**Comment by philiptaron** [comment]
Is there a tracking issue for removing all the now-obviated `lib.mdDoc` calls?

**Comment by philiptaron** [comment]
> Is there a tracking issue for removing all the now-obviated `lib.mdDoc` calls?

I made one. 
- https://github.com/NixOS/nixpkgs/issues/300735

File: nixos/modules/misc/documentation.nix [hunk 2]
```diff
@@ -51,12 +76,11 @@ let
           (name: value:
             let
               wholeName = "${namePrefix}.${name}";
-              guard = lib.warn "Attempt to evaluate package ${wholeName} in option documentation; this is not supported and will eventually be an error. Use `mkPackageOption{,MD}` or `literalExpression` instead.";
+              guard = warn "Attempt to evaluate package ${wholeName} in option documentation; this is not supported and will eventually be an error. Use `mkPackageOption{,MD}` or `literalExpression` instead.";
             in if isAttrs value then
               scrubDerivations wholeName value
               // optionalAttrs (isDerivation value) {
                 outPath = guard "\${${wholeName}}";
-                drvPath = guard drvPath;
```
Comment (DanielSidhion) [inline nixos/modules/misc/documentation.nix:59]: Why was this line removed?

Comment (philiptaron) [reply nixos/modules/misc/documentation.nix:59]: It doesn't eval; there's no drvPath in scope. It was masked by the top-level with.

Comment (philiptaron) [reply nixos/modules/misc/documentation.nix:59]: I noted this in the PR head matter.

Comment (DanielSidhion) [reply nixos/modules/misc/documentation.nix:59]: Ah, it all makes sense now. I didn't understand that line when I first read it in the description.

**Review by pennae — DISMISSED** [review]

File: nixos/modules/misc/documentation.nix [hunk 3]
```diff
@@ -51,12 +76,11 @@ let
           (name: value:
             let
               wholeName = "${namePrefix}.${name}";
-              guard = lib.warn "Attempt to evaluate package ${wholeName} in option documentation; this is not supported and will eventually be an error. Use `mkPackageOption{,MD}` or `literalExpression` instead.";
+              guard = warn "Attempt to evaluate package ${wholeName} in option documentation; this is not supported and will eventually be an error. Use `mkPackageOption{,MD}` or `literalExpression` instead.";
             in if isAttrs value then
               scrubDerivations wholeName value
               // optionalAttrs (isDerivation value) {
                 outPath = guard "\${${wholeName}}";
```
Comment (DanielSidhion) [inline nixos/modules/misc/documentation.nix:83]: Can't make a suggestion on the deleted line, so on a new comment. Let's fix the bug:

```suggestion
                outPath = guard "\${${wholeName}}";
                drvPath = guard value.drvPath;
```

File: nixos/modules/misc/documentation.nix (hunk 2 above)
Comment (philiptaron) [reply nixos/modules/misc/documentation.nix:59]: I'd be glad to, but I don't understand how it's a bug if the code wasn't ever evaluated in the first place. Doesn't that just meet the definition of dead code?

Definitely this code meant to do something, and I'd be happy to help make it do whatever it was meant to do.

Comment (DanielSidhion) [reply nixos/modules/misc/documentation.nix:59]: I was writing the suggestion! :) Couldn't suggest a fix because GH doesn't like suggestion on removed lines. Here it is: https://github.com/NixOS/nixpkgs/pull/299650#discussion_r1546992510

Comment (DanielSidhion) [reply nixos/modules/misc/documentation.nix:59]: To answer the question here since it got asked here: this is not dead code because the purpose of the code is to ensure that it will not be used - hence the `lib.warn` line. It exists to let people know what to do in case they eventually hit this code by accident. If the code isn't ever evaluated, it's doing its job.

File: nixos/modules/misc/documentation.nix (hunk 3 above)
Comment (philiptaron) [reply nixos/modules/misc/documentation.nix:83]: Wonderful, I’ll do that.

**Comment by philiptaron** [comment]
Updated. Thank you all for your wisdom.

**Review by DanielSidhion — APPROVED** [review]
LGTM

**Comment by infinisil** [comment]
> Spot the bug with drvPath = guard drvPath; though. Never evaled, and under a global with, it's not an error.

Damn nice one!

**Review by infinisil — APPROVED** [review]
